### PR TITLE
Alt-branch-1 ready to be merged into main

### PR DESCRIPTION
**Pull request by Fernando-Iscar**
Hi guys,

As discussed in the last meeting, we will continue working with unit testing in the new branch `test-development-branch`.

Thanks!

@DiSanchz

### CODE REVIEW

File: pyproject.toml
Comment (narendramukherjee) [inline pyproject.toml]: @Fernando-Iscar @DiSanchz - since this is to do with the authors of the codebase rather than the paper, you guys should add your names as well

File: snpe_reputation_systems/poetry.lock
Comment (narendramukherjee) [inline snpe_reputation_systems/poetry.lock:1]: Is the poetry lock file usually committed in a repository as well? Does it change every time you run poetry?

File: snpe_reputation_systems/snpe_reputation_systems/utils/functions.py
Comment (narendramukherjee) [inline snpe_reputation_systems/snpe_reputation_systems/utils/functions.py]: `mypy` has raised an error here as the return type of this function is not actually `None` - it is actually taking a Linux command `cmd`, running it and "yielding" the output of running this command on line 19 (`yield stdout_line`). If you don't do this, the output of a command run from Python in a subprocess, like here, will only be seen after the command is completed. We want to use this to run the starspace embeddings fitting command, which will take a while to run - if you don't have this functionality, you will not know how much Starspace has progressed.

So here the return type is a Python `Generator`, or if only yielding and nothing else, `Iterator` - generators are special Python objects which `yield` things: for eg., the `range` function is also a Generator. Try looking at the type of `range(5)` yourself.

So here, the fix is to have a return type of `Iterator[str]` following the instructions here: https://stackoverflow.com/questions/43658999/what-is-the-return-type-hint-of-a-generator-function
You need to also `from typing import Iterator` of course

File: snpe_reputation_systems/snpe_reputation_systems/utils/embedding_nets.py
Comment (narendramukherjee) [inline snpe_reputation_systems/snpe_reputation_systems/utils/embedding_nets.py:5]: `mypy` asks to `from typing import List` here. Check log of pipeline

**Review by narendramukherjee — CHANGES_REQUESTED** [review]
Suggested the fixes needed for mypy

File: pyproject.toml
Comment (Fernando-Iscar) [reply pyproject.toml]: Names added, thanks! 
@narendramukherjee

File: snpe_reputation_systems/poetry.lock
Comment (DiSanchz) [reply snpe_reputation_systems/poetry.lock:1]: **Regarding the first question:**

This particular "poetry.lock" file was generated and committed at the time when the functioning of poetry was being tested for the first time. At that point, if I remember correctly, I was still assuming (incorrectly) that poetry would play some role in dependency management in the project.

Having said that, the "poetry.lock" file is indeed usually committed in repositories as a tool to ensure that all collaborators work under the same version of the relevant dependencies. However, as far as I have read, this holds true mostly when at least one of the following conditions is met:

1. The nature of the project makes the inclusion of the lock file particularly interesting or advisable
2. Poetry is the primary tool for dependency management

In our case, none of these are met given that:

1. End users of a Python package interact with it via a package installer, which uses the dependency specifications in pyproject.toml, not "poetry.lock". Therefore, the "poetry.lock" file doesn't influence the package installation for users, making it less relevant to include in a package's repository. Moreover, packages typically aim at being compatible with a range of dependencies, rather than the specific versions "locked" at a given time, which also diminishes the relevance of "poetry.lock" as part of the repository.
2. Poetry is not being used as our primary tool for dependency management.

The "poetry.lock" can also contribute to a project's reproducibility to an extent, however, I understand we are already addressing this with the use of docker containers. Thus the benefits in this sense would be rather limited considering that the yet-to-be-added dockerfile(s) for the new repo contain appropriate instructions on the range of dependency versions within which we expect it to be working.

**Regarding the second question:**

A poetry.lock file is generated whenever the `poetry install` command is run and poetry could not find an existing "poetry.lock" within its context, based on the information on dependencies included in the "pyproject.toml". Moreover, (according to poetry's documentation) it is also updated/generated whenever the project's dependencies are modified through poetry commands (mainly `poetry remove`, `poetry add`, `poetry remove`, which also affect the "pyproject.toml"). So, in short, it does not necessarily change every time the repository is "packaged" as long as the dependency information inside the "pyproject.toml" has not been modified as well.

**As a personal conclusion/opinion**

Whether to commit the "poetry.lock" largely depends on the particular situation. Given the factors considered above and our particular use context, I lean to think that the reasons for not committing it outweigh those others in favor of doing so. If you agree, I would remove it from the branch before merging.

**Sources**

For further insight, there are some other factors at play regarding the tradeoffs of including or not the "poetry.lock" file (which I have considered less relevant). These sources mention some of them, alongside those already discussed above.

https://python-poetry.org/docs/basic-usage/#:~:text=Committing%20your%20poetry.lock%20file%20to%20version%20control,-As%20an%20application&text=Application%20developers%20commit%20poetry.lock,dependencies%20that%20you%20are%20using. [Section: "Committing your poetry.lock file to version control"]

https://github.com/python-poetry/poetry/issues/7488

**Comment by DiSanchz** [comment]
As an additional relevant note on the PR, the last commits by Fernando include a modification of the workflow which now will be marked as failed if any formatting errors are found in the code as we discussed during the last meeting.

File: snpe_reputation_systems/poetry.lock
Comment (narendramukherjee) [reply snpe_reputation_systems/poetry.lock:1]: Sounds good @DiSanchz - let's remove the lock file then. Also, at the next meeting, let's briefly go over what's doing dependency management right now and how the library is being packaged

File: snpe_reputation_systems/snpe_reputation_systems/utils/embedding_nets.py
Comment (Fernando-Iscar) [reply snpe_reputation_systems/snpe_reputation_systems/utils/embedding_nets.py:5]: Ok, now it's working 👍

File: snpe_reputation_systems/snpe_reputation_systems/utils/functions.py
Comment (Fernando-Iscar) [reply snpe_reputation_systems/snpe_reputation_systems/utils/functions.py]: Ok, we implemented these changes 👍 👍

**Comment by Fernando-Iscar** [comment]
Hi @narendramukherjee @shrabasteebanerjee ,

As you may see below, we solved the issues with the workflow and now all checks are passing. Hence, in that regard, the PR should be ready.

Let us know if you have any further suggestions

**Review by narendramukherjee — APPROVED** [review]

**Comment by narendramukherjee** [comment]
I have approved the changes, can @shrabasteebanerjee approve these changes as well? We need 2 approvals for merging

@DiSanchz @Fernando-Iscar When the merge button is available when 2 approvals have been obtained, be careful that you select "squash commits while merging" by clicking on the downward triangle next to the merge button. Then the MR makes only 1 merge commit and the commit history is cleaner

**Review by shrabasteebanerjee — APPROVED** [review]

**Comment by DiSanchz** [comment]
We got it! Let's merge it!

**Pull request merged**
